src/main/ruby/truffleruby/core/string.rb

```diff
@@ -1097,6 +1097,82 @@ def rindex(sub, finish = undefined)
     nil
   end
 
+  def byteindex(str, start = undefined)
+    is_regex_pattern = Primitive.is_a?(str, Regexp)
+
+    if Primitive.undefined?(start)
+      start = 0
+    else
+      start = Primitive.rb_to_int(start)
+
+      start += bytesize if start < 0
+      if start < 0 || start > bytesize
+        if is_regex_pattern
+          Primitive.regexp_last_match_set(Primitive.caller_special_variables, nil)
+        end
+
+        return nil
+      end
+    end
+
+    unless Truffle::StringOperations.on_codepoint_boundary?(self, start)
+      raise IndexError, "offset #{start} does not land on character boundary"
+    end
+
+    if is_regex_pattern
+      Primitive.encoding_ensure_compatible(self, str)
+
+      match = Truffle::RegexpOperations.match_from(str, self, start)
+      Primitive.regexp_last_match_set(Primitive.caller_special_variables, match)
+      return match ? Primitive.character_index_to_byte_index(self, match.begin(0)) : nil
+    end
+
+    str = StringValue(str)
+    return start if str.empty?
+    return nil if start + str.bytesize > bytesize
+
+    enc = Primitive.encoding_ensure_compatible_str(self, str)
+    Primitive.string_byte_index(self, str, enc, start)
+  end
+
+  def byterindex(str, finish = undefined)
+    if Primitive.undefined?(finish)
+      finish = bytesize
+    else
+      finish = Primitive.rb_to_int(finish)
+      finish += bytesize if finish < 0
+      return nil if finish < 0
+
+      finish = bytesize if finish > bytesize
+    end
+
+    unless Truffle::StringOperations.on_codepoint_boundary?(self, finish)
+      raise IndexError, "offset #{finish} does not land on character boundary"
+    end
+
+    if Primitive.is_a?(str, Regexp)
+      Primitive.encoding_ensure_compatible(self, str)
+
+      match = Truffle::RegexpOperations.search_region(str, self, 0, finish, false, true)
+      Primitive.regexp_last_match_set(Primitive.caller_special_variables, match)
+      return match ? Primitive.character_index_to_byte_index(self, match.begin(0)) : nil
+    end
+
+    str = StringValue(str)
+    return finish if str.empty?
+    return nil if str.bytesize > bytesize
+
+    # Add `str.size` worth of bytes to `finish` to compensate for `LastByteIndexOfStringNode` which does a reverse
+    # search including the full pattern length (as start pos).
+    finish_adjusted = Primitive.byte_index_to_character_index(self, finish)
+    finish_adjusted += str.size
+    finish_adjusted = size if finish_adjusted > size
+    finish = Primitive.character_index_to_byte_index(self, finish_adjusted)
```

Comment:
Can't this be something like `finish += str.bytesize` ?
I don't understand why this is necessary or what it does though.

Reply:
I guess I could have made the comment above better. Let's say the call is

```ruby
("x" * 10).byterindex("xxx", 5)
```

Ruby will start the lookup matching the pattern on index 5:

```
xxxxxxxxxx
     xxx
```

`StringByteReverseIndexNode` on the other hand is matching the pattern at the end (non inclusive) on index 2:

```
xxxxxxxxxx
  xxx
```

What makes this offset adjustment non trivial is the difference in encoding. Since `str` might have different char length, `finish += str.bytesize` can fall on a non codepoint boundary. Conceptually we need to adjust `str.size` characters in bytes. Is there a better way to do this?

Reply:
> Ruby will start the lookup matching the pattern on index 5:

These docs are pretty confusing: https://docs.ruby-lang.org/en/master/String.html#method-i-byterindex
> Integer argument offset, if given and non-negative, specifies the maximum starting byte-based position in the string to _end_ the search:

But indeed:
```
> ("x" * 10).byterindex("xxx", 10)
=> 7
> ("x" * 10).byterindex("xxx", 9)
=> 7
> ("x" * 10).byterindex("xxx", 8)
=> 7
> ("x" * 10).byterindex("xxx", 7)
=> 7
> ("x" * 10).byterindex("xxx", 6)
=> 6
> ("x" * 10).byterindex("xxx", 5)
=> 5
> ("x" * 10).byterindex("xxx", 4)
=> 4
```

So conceptually it's like a maximum limit to the returned value.

In my understanding, adding `finish = MIN(finish + str.bytesize, self.bytesize)` is correct then.

> can fall on a non codepoint boundary.

Why does it matter? I think if TruffleString can handle it we can just ignore that.
If it can't we could keep increasing finish by 1 until that position is a character head.

Reply:
> Why does it matter? I think if TruffleString can handle it we can just ignore that.

I had the assumption it matters. If we don't care and the code neither (tested, it doesn't), than I guess it's fine. Fixed.